course/include/insect.h

```diff
@@ -0,0 +1,54 @@
+#ifndef INSECT_H
+#define	INSECT_H
+#include <cstdint>
+#include <functional>
+
+#include "cell_mock.h"
+#include "insects_metrics.h"
+
+namespace ekumen {
+namespace simulation {
+
+class Cell;
+
+// Abstract class that implements the common behaviour of every simulated insect,
+// and defines the methods that must be implemented by every insect.
+// We need the enable_shared_from_this ineritance to pass the instance's reference when moving
+// to a new cell.
+class Insect {
+    public:
+        // Simulates a round.
+        virtual void RunRound() = 0;
+        void Die();
+        void SetRoundResultsCallback(std::function<void(InsectCallbackMetrics&)> callback);
+
+    protected:
+        // Making these virtual allows to easily change the insect behaviour if
+        // needed, or have additional functionalities besides the basic ones.
+        virtual void Breed();
+        virtual void Move();
+        virtual void Eat();
+        virtual InsectType GetInsectType() = 0;
```

Comment:
Mmm... why? I mean, the only thing that the environment should tell the insect is to simulate... it shouldn't care about how they do it. At least from my POV.

Breed(), Move() and Eat() are the internal mechanism of how the insects implement the RunRound() method.

Reply:
I guess the idea is that some general method in Insect needs to know which actual insect this instance is. Otherwise the specialized insect type already knows what it is.

Reply:
You're right. the GetInsectType() method should be public.